akka-http/src/main/scala/akka/http/scaladsl/server/directives/MdcLoggingDirectives.scala

```diff
@@ -0,0 +1,70 @@
+package akka.http.scaladsl.server.directives
+
+import akka.event._
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server._
+
+/**
+ * @groupname mdc Mdc logging directives
+ * @groupprio mdc 240
+ */
+trait MdcLoggingDirectives {
+
+  import MdcLoggingDirectives.createDiagnosticMarkerLoggingAdapter
+
+  def withMarkerLoggingAdapter: Directive1[MarkerLoggingAdapter] = {
+    createDiagnosticMarkerLoggingAdapter(Map.empty)
+      .flatMap { dmbla: DiagnosticMarkerBusLoggingAdapter =>
+        mapRequestContext { ctx =>
+          ctx.withLog(dmbla)
+        } & provide {
+          dmbla: MarkerLoggingAdapter
+        }
+      }
+  }
+
+  def withMdcEntries(mdc: (String, Any)*): Directive0 =
+    createDiagnosticMarkerLoggingAdapter(mdc.toMap)
+      .flatMap { dmbla: DiagnosticMarkerBusLoggingAdapter =>
+        mapRequestContext { ctx =>
+          ctx.withLog(dmbla)
+        }
+      }
+
+  def withMdcEntry(key: String, value: Any): Directive0 =
+    withMdcEntries((key, value))
+}
+
+object MdcLoggingDirectives extends MdcLoggingDirectives {
+  private def createDiagnosticMarkerLoggingAdapter(entries: Map[String, Any]): Directive1[DiagnosticMarkerBusLoggingAdapter] =
+    extractActorSystem
+      .flatMap { sys =>
+        extractRequestContext
+          .flatMap { ctx =>
+            provide {
+              // Convert the existing ctx.log LoggingAdapter to a DiagnosticMarkerBusLoggingAdapter.
+              // This requires various amounts of effort depending on the type of the current LoggingAdapter.
+              // In all cases, we create a _new_ DiagnosticMarkerBusLoggingAdapter in order to avoid race conditions,
+              // where multiple requests could add/remove entries from the same MDC map.
+              ctx.log match {
+                case existingAdapter: DiagnosticMarkerBusLoggingAdapter =>
+                  import existingAdapter._
+                  val newAdapter = new DiagnosticMarkerBusLoggingAdapter(bus, logSource, logClass, loggingFilterWithMarker)
+                  newAdapter.mdc(existingAdapter.mdc ++ entries)
```

Comment:
This part is concerning me a bit. Assume, for whatever reason, the logger is always already a singleton instance of `DiagnosticMarkerBusLoggingAdapter` (maybe defined once in `main` and provided to every request via `mapRequest` or similar). Then it seems the mdc map will never be cleared out, leading to larger and larger maps, i.e., a memory leak. There probably needs to be some case where we clear out this map.

Reply:
The original logging adapter is not mutated here though, so when `newAdapter` is GC:d the mdc map is as well. (If we would set the combined mdc on the original adapter without cleanup that would be a leak)

Reply:
I think you're right. We're always copying the existing MDC entries and then adding values. We're not just perpetually adding to the original map.